src/evaluators/UnaryExpression.js

```diff
@@ -29,13 +31,122 @@ import {
 }
 
+function evaluateDeleteOperation(expr: Value | Reference, realm: Realm) {
+  // ECMA262 12.5.3.2
+
+  // 1. Let ref be the result of evaluating UnaryExpression.
+  let ref = expr;
+
+  // 2. ReturnIfAbrupt(ref).
+
+  // 3. If Type(ref) is not Reference, return true.
+  if (!(ref instanceof Reference)) return realm.intrinsics.true;
+
+  // 4. If IsUnresolvableReference(ref) is true, then
+  if (Environment.IsUnresolvableReference(realm, ref)) {
+    // a. Assert: IsStrictReference(ref) is false.
+    invariant(!Environment.IsStrictReference(realm, ref), "did not expect a strict reference");
+
+    // b. Return true.
+    return realm.intrinsics.true;
+  }
+
+  // 5. If IsPropertyReference(ref) is true, then
+  if (Environment.IsPropertyReference(realm, ref)) {
+    // a. If IsSuperReference(ref) is true, throw a ReferenceError exception.
+    if (Environment.IsSuperReference(realm, ref)) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.ReferenceError);
+    }
+
+    // b. Let baseObj be ! ToObject(GetBase(ref)).
+    let base = Environment.GetBase(realm, ref);
+    // Constructing the reference checks that base is coercible to an object hence
+    invariant(base instanceof ConcreteValue || base instanceof AbstractObjectValue);
+    let baseObj = base instanceof ConcreteValue ? To.ToObject(realm, base) : base;
+
+    // c. Let deleteStatus be ? baseObj.[[Delete]](GetReferencedName(ref)).
+    let deleteStatus = baseObj.$Delete(Environment.GetReferencedName(realm, ref));
+
+    // d. If deleteStatus is false and IsStrictReference(ref) is true, throw a TypeError exception.
+    if (!deleteStatus && Environment.IsStrictReference(realm, ref)) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.TypeError);
+    }
+
+    // e. Return deleteStatus.
+    return new BooleanValue(realm, deleteStatus);
+  }
+
+  // 6. Else ref is a Reference to an Environment Record binding,
+  // a. Let bindings be GetBase(ref).
+  let bindings = Environment.GetBase(realm, ref);
+  invariant(bindings instanceof EnvironmentRecord);
+
+  // b. Return ? bindings.DeleteBinding(GetReferencedName(ref)).
+  let referencedName = Environment.GetReferencedName(realm, ref);
+  invariant(typeof referencedName === "string");
+  return new BooleanValue(realm, bindings.DeleteBinding(referencedName));
+}
+
+function generateRuntimeCall(ref: Reference | Value, ast: BabelNodeUnaryExpression, strictCode: boolean, realm: Realm) {
+  invariant(ref instanceof Reference);
+  let baseValue = Environment.GetBase(realm, ref);
+  invariant(baseValue instanceof Value);
+  let propertyName = Environment.GetReferencedName(realm, ref);
+  invariant(typeof propertyName === "string");
+
+  if (!baseValue.isSimpleObject()) {
+    Leak.leakValue(realm, baseValue, ast.loc);
+  }
+  return AbstractValue.createTemporalFromBuildFunction(realm, Value, [baseValue], nodes => {
+    let arg;
+    if (typeof propertyName === "string") {
+      arg = t.isValidIdentifier(propertyName)
```

Comment:
Consider using or generalizing the `Generator`'s `getAsPropertyNameExpression`.

Reply:
I'd prefer to do this in a follow-up PR as there are many places the pattern I used above exists. They should all be change in one PR for consistency.